[AutoFill Debugging] Add an option to limit text extraction output to visible text

whsieh opened the pull request:
#### 9bce2e4947440a85fe148568a6e0f20184bd188d
<pre>
[AutoFill Debugging] Add an option to limit text extraction output to visible text
<a href="https://bugs.webkit.org/show_bug.cgi?id=302053">https://bugs.webkit.org/show_bug.cgi?id=302053</a>
<a href="https://rdar.apple.com/161663236">rdar://161663236</a>

Reviewed by Abrar Rahman Protyasha.

Add support for `+[_WKTextExtractionConfiguration configurationForVisibleTextOnly]`, which returns
a text extraction configuration that only extracts visible text from the DOM (applying default text
filtering heuristics, but ignoring all other DOM attributes and item types).

Test: TextExtractionTests.VisibleTextOnly

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::addResult):
(WebKit::TextExtractionAggregator::includeRects const):
(WebKit::TextExtractionAggregator::includeURLs const):
(WebKit::TextExtractionAggregator::onlyIncludeText const):
(WebKit::TextExtractionAggregator::addLineForNativeMenuItemsIfNeeded):
(WebKit::addPartsForText):
(WebKit::addPartsForItem):
(WebKit::addTextRepresentationRecursive):

Teach `TextExtractionAggregator` (and related code) to honor the new flag by only emitting output
for text items (passing each piece of text through the normal filtering mechanisms, including OCR).

* Source/WebKit/Shared/TextExtractionToStringConversion.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _debugTextWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration init]):
(+[_WKTextExtractionConfiguration configurationForVisibleTextOnly]):
(-[_WKTextExtractionConfiguration _initForOnlyVisibleText:]):
(-[_WKTextExtractionConfiguration _logInvalidTextOnlyConfiguration:]):
(-[_WKTextExtractionConfiguration setIncludeURLs:]):
(-[_WKTextExtractionConfiguration setIncludeRects:]):
(-[_WKTextExtractionConfiguration setIncludeNodeIdentifiers:]):
(-[_WKTextExtractionConfiguration setIncludeEventListeners:]):
(-[_WKTextExtractionConfiguration setIncludeAccessibilityAttributes:]):
(-[_WKTextExtractionConfiguration setIncludeTextInAutoFilledControls:]):

Adjust some of these setters to log a warning message if a client attempts to set `include{…}` to
`YES`, on a configuration returned by `+configurationForVisibleTextOnly`. These flags ultimately
have no effect here, since the result of `+configurationForVisibleTextOnly` will ensure that we
only ever dump text content (not attributes, geometry, identifiers, or event listeners).

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, VisibleTextOnly)):

Add a new API test to exercise the new configuration.

Canonical link: <a href="https://commits.webkit.org/302652@main">https://commits.webkit.org/302652@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/756f683663f9a63d606d36016a4a695f5361e95b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129759 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40615 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137148 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81228 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d213538d-0170-4726-bbad-6c379953f7fd) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131630 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2010 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1909 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98851 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66672 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f44d9dd3-7865-475a-bc9e-0fbe7a349cdf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132706 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1528 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116218 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79530 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/21d81b5a-516d-4c69-9a48-a1a9924cb529) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1444 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34350 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80421 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109930 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139631 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1814 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1697 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107357 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1859 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112566 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107232 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27304 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1470 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31056 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54571 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1887 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65256 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1701 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1736 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1808 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
```diff
@@ -89,6 +101,59 @@ - (void)setReplacementStrings:(NSDictionary<NSString *, NSString *> *)replacemen
+{
+    if (_onlyIncludeVisibleText && value)
+        return [self _logInvalidTextOnlyConfiguration:__PRETTY_FUNCTION__];
+
```
optional, did you consider making this a macro? Something like `BAIL_IF_TEXT_ONLY_CONFIGURATION(value)`?

whsieh replied:
Oh, that would probably work better since I don't need to pass the function name around — will make that change.

Thanks for the review!

whsieh committed:
[AutoFill Debugging] Add an option to limit text extraction output to visible text
https://bugs.webkit.org/show_bug.cgi?id=302053
rdar://161663236

Reviewed by Abrar Rahman Protyasha.

Add support for `+[_WKTextExtractionConfiguration configurationForVisibleTextOnly]`, which returns
a text extraction configuration that only extracts visible text from the DOM (applying default text
filtering heuristics, but ignoring all other DOM attributes and item types).

Test: TextExtractionTests.VisibleTextOnly

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::addResult):
(WebKit::TextExtractionAggregator::includeRects const):
(WebKit::TextExtractionAggregator::includeURLs const):
(WebKit::TextExtractionAggregator::onlyIncludeText const):
(WebKit::TextExtractionAggregator::addLineForNativeMenuItemsIfNeeded):
(WebKit::addPartsForText):
(WebKit::addPartsForItem):
(WebKit::addTextRepresentationRecursive):

Teach `TextExtractionAggregator` (and related code) to honor the new flag by only emitting output
for text items (passing each piece of text through the normal filtering mechanisms, including OCR).

* Source/WebKit/Shared/TextExtractionToStringConversion.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _debugTextWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration init]):
(+[_WKTextExtractionConfiguration configurationForVisibleTextOnly]):
(-[_WKTextExtractionConfiguration _initForOnlyVisibleText:]):
(-[_WKTextExtractionConfiguration _logInvalidTextOnlyConfiguration:]):
(-[_WKTextExtractionConfiguration setIncludeURLs:]):
(-[_WKTextExtractionConfiguration setIncludeRects:]):
(-[_WKTextExtractionConfiguration setIncludeNodeIdentifiers:]):
(-[_WKTextExtractionConfiguration setIncludeEventListeners:]):
(-[_WKTextExtractionConfiguration setIncludeAccessibilityAttributes:]):
(-[_WKTextExtractionConfiguration setIncludeTextInAutoFilledControls:]):

Adjust some of these setters to log a warning message if a client attempts to set `include{…}` to
`YES`, on a configuration returned by `+configurationForVisibleTextOnly`. These flags ultimately
have no effect here, since the result of `+configurationForVisibleTextOnly` will ensure that we
only ever dump text content (not attributes, geometry, identifiers, or event listeners).

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, VisibleTextOnly)):

Add a new API test to exercise the new configuration.

Canonical link: https://commits.webkit.org/302652@main
